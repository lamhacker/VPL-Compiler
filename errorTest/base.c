#include "base.h"

float* mallocVec(int num)
{
    float* vec = malloc(sizeof(float) * num + SSE_ALIGN);
	vec = (float *) align(vec, SSE_ALIGN);
	return vec;
}

void freeVec(float* vec)
{
    free(vec);
}

void setValue(float* vec, float val, int num)
{
    for (int i = 0; i < num; i++)
    {
        vec[i] = val;
    }
}

void printVec(char* rep, float* vec, int num)
{
	printf("%s: ", rep);
    for (int i = 0; i < num; i++)
    {   
        printf("%f ", vec[i]);
    }
	printf("\n");
}


