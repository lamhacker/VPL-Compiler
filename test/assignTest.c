#include "base.h"
#define NUM (4)

extern void assign(long, float*);
extern void assignTwo(long, float*, float*);
extern void assignThree(long, float*, float*);
extern void assignFour(long, float*, float*);
extern void assignFive(long, float*, float*);
extern void assignSix(long, float*, float*, float*);
extern void assignSeven(long, float*, float*);



int main(void)
{
    float* a = mallocVec(NUM);
	float* b = mallocVec(NUM);
	float* c = mallocVec(NUM);
	
	printf("assign: \n");
    setValue(a, 0, NUM);
    assign(NUM, a);
	printVec("a", a, NUM);
	
	printf("assignTwo: \n");
	setValue(a, 0, NUM);
	setValue(b, 0, NUM);
	assignTwo(NUM, a, b);
	printVec("a", a, NUM);
	printVec("b", b, NUM);
	
	printf("assignThree: \n");
	setValue(a, 1, NUM);
	setValue(b, 3, NUM);
	assignThree(NUM, a, b);
	printVec("a", a, NUM);
	printVec("b", b, NUM);
	
	printf("assignFour: \n");
	setValue(a, 1, NUM);
	setValue(b, 3, NUM);
	assignFour(NUM, a, b);
	printVec("a", a, NUM);
	printVec("b", b, NUM);
	
	printf("assignFive: \n");
	setValue(a, 1, NUM);
	setValue(b, 3, NUM);
	assignFive(NUM, a, b);
	printVec("a", a, NUM);
	printVec("b", b, NUM);
	
	printf("assignSix: \n");
	setValue(a, 1, NUM);
	setValue(b, 3, NUM);
	setValue(c, 4, NUM);
	assignSix(NUM, a, b, c);
	printVec("a", a, NUM);
	printVec("b", b, NUM);
	printVec("c", c, NUM);
	
	printf("assignSeven: \n");
	setValue(a, 1, NUM);
	setValue(b, 3, NUM);
	assignSeven(NUM, a, b);
	printVec("a", a, NUM);
	printVec("b", b, NUM);
	
	freeVec(a);
	freeVec(b);
	freeVec(c);
}
