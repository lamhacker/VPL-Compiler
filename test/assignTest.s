
.text
.global assign
.type assign, @function
.p2align 4,,15

assign:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of 1.0 into %rax
    movq $.const1.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end0

.loop_begin0:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin0
.loop_end0:


    popq    %rbx
    leave
    ret

.text
.global assignTwo
.type assignTwo, @function
.p2align 4,,15

assignTwo:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of 2.0 into %rax
    movq $.const2.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end1

.loop_begin1:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin1
.loop_end1:

    # load source address into %rax
    
    # place address of 4.0 into %rax
    movq $.const4.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rdx, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end2

.loop_begin2:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin2
.loop_end2:


    popq    %rbx
    leave
    ret

.text
.global assignThree
.type assignThree, @function
.p2align 4,,15

assignThree:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rdx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end3

.loop_begin3:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin3
.loop_end3:


    popq    %rbx
    leave
    ret

.text
.global assignFour
.type assignFour, @function
.p2align 4,,15

assignFour:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rdx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end4

.loop_begin4:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin4
.loop_end4:

    # load source address into %rax
    
    # place address of 2.0 into %rax
    movq $.const2.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end5

.loop_begin5:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin5
.loop_end5:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rsi, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rdx, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end6

.loop_begin6:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin6
.loop_end6:


    popq    %rbx
    leave
    ret

.text
.global assignFive
.type assignFive, @function
.p2align 4,,15

assignFive:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rsi, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rdx, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end7

.loop_begin7:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin7
.loop_end7:

    # load source address into %rax
    
    # place address of 2.0 into %rax
    movq $.const2.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end8

.loop_begin8:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin8
.loop_end8:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rdx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end9

.loop_begin9:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin9
.loop_end9:


    popq    %rbx
    leave
    ret

.text
.global assignSix
.type assignSix, @function
.p2align 4,,15

assignSix:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 0 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $0, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rdx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end10

.loop_begin10:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin10
.loop_end10:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rcx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rdx, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end11

.loop_begin11:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin11
.loop_end11:

    # load source address into %rax
    
    # place address of 1.0 into %rax
    movq $.const1.0, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rcx, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end12

.loop_begin12:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %r10

    decq %rbx
    jnz .loop_begin12
.loop_end12:


    popq    %rbx
    leave
    ret

.text
.global assignSeven
.type assignSeven, @function
.p2align 4,,15

assignSeven:
    pushq   %rbp
    movq    %rsp, %rbp
    pushq   %rbx

    # allocate 1 local variables
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $1, %rax, %rax
    subq %rax, %rsp
    andq $-16, %rsp

    
    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rsi, %rax


    # load destination address into %r10
    
    # place address of 1th local variable into %r10
    movq %rdi, %r10
    imulq $4, %r10, %r10
    addq $16, %r10
    imulq $1, %r10, %r10
    subq %rbp, %r10
    negq %r10
    andq $-16, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end13

.loop_begin13:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin13
.loop_end13:

    # load source address into %rax
    
    # place address of parameter into %rax
    movq %rdx, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rsi, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end14

.loop_begin14:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin14
.loop_end14:

    # load source address into %rax
    
    # place address of 1th local variable into %rax
    movq %rdi, %rax
    imulq $4, %rax, %rax
    addq $16, %rax
    imulq $1, %rax, %rax
    subq %rbp, %rax
    negq %rax
    andq $-16, %rax


    # load destination address into %r10
    
    # place address of parameter into %r10
    movq %rdx, %r10


    movq %rdi, %rbx                    
    shrq $2, %rbx                      
                                       
    jz .loop_end15

.loop_begin15:
    movaps (%rax), %xmm0        # load source into %xmm0
    movaps %xmm0, (%r10)        # store %xmm0


    addq $16, %rax             

    addq $16, %r10

    decq %rbx
    jnz .loop_begin15
.loop_end15:


    popq    %rbx
    leave
    ret

.data 
.align 16 
.const1.0:
    .float 1.0 
    .float 1.0 
    .float 1.0 
    .float 1.0

.data 
.align 16 
.const2.0:
    .float 2.0 
    .float 2.0 
    .float 2.0 
    .float 2.0

.data 
.align 16 
.const4.0:
    .float 4.0 
    .float 4.0 
    .float 4.0 
    .float 4.0

