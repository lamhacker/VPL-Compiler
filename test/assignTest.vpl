func assign(a)
	a = 1;
end

func assignTwo(a, b)
	a = 2;
	b = 4;
end

func assignThree(a, b)
	a = b;
end

func assignFour(a, b)
	a = b;
	a = 2;
	b = a;
end

func assignFive(a, b)
	b = a;
	a = 2;
	a = b;
end

func assignSix(a, b, c)
	a = b;
	b = c;
	c = 1;
end

func assignSeven(a, b)
	var c;
	c = a;
	a = b;
	b = c;
end
