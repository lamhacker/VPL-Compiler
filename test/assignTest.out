assign: 
a: 1.000000 1.000000 1.000000 1.000000 
assignTwo: 
a: 2.000000 2.000000 2.000000 2.000000 
b: 4.000000 4.000000 4.000000 4.000000 
assignThree: 
a: 3.000000 3.000000 3.000000 3.000000 
b: 3.000000 3.000000 3.000000 3.000000 
assignFour: 
a: 2.000000 2.000000 2.000000 2.000000 
b: 2.000000 2.000000 2.000000 2.000000 
assignFive: 
a: 1.000000 1.000000 1.000000 1.000000 
b: 1.000000 1.000000 1.000000 1.000000 
assignSix: 
a: 3.000000 3.000000 3.000000 3.000000 
b: 4.000000 4.000000 4.000000 4.000000 
c: 1.000000 1.000000 1.000000 1.000000 
assignSeven: 
a: 3.000000 3.000000 3.000000 3.000000 
b: 1.000000 1.000000 1.000000 1.000000 
